File: plugins/doc.py
import re

Beeld = None
def beeld_plugin_doc(beelder):
    global Beeld
    Beeld = beelder.getClass() #BeeldClass
    beelder.addAction('doc', beeld_plugin_action_doc)

def beeld_plugin_action_doc(evt):
    global Beeld
    params = evt.data
    options = params.options
    data = params.data
    current = params.current
    doc = current.action_cfg
    if doc and ('output' in doc):

        doc['output'] = Beeld.Utils.xpresion(doc['output'], evt) # parse xpresion if any
        docFile = Beeld.Utils.get_real_path(Beeld.Utils.evaluate(doc['output'], {}), options.basePath)
        startDoc = doc['startdoc']
        endDoc = doc['enddoc']

        _trim = None
        _trimlen = 0
        isRegex = 0

        sep = doc['separator'] if 'separator' in doc else "\n\n"

        if 'trim' in doc:
            _trim = Beeld.Utils.regex(doc['trim'], evt)
            if _trim is False:
                _trim = doc['trim']
                _trimlen = len(_trim)
                isRegex = 0
            else:
                isRegex = 1


        docs = []

        # extract doc blocks
        blocks = data.src.split(startDoc)
        for b in blocks:
            tmp = b.split(endDoc)
            if len(tmp)>1: docs.append(tmp[0])
        blocks = None

        # trim start of each doc block line
        if _trim:
            for i in range(len(docs)-1):
                tmp = docs[i].split("\n")

                for j in range(len(tmp)-1):
                    if len(tmp[j])>0:
                        if isRegex:
                            tmp[j] = re.sub(_trim, '', tmp[j])
                        elif tmp[j].startswith(_trim):
                            tmp[j] = tmp[j][_trimlen:]

                docs[i] = "\n".join(tmp)
        Beeld.Utils.write(docFile, sep.join(docs), options.encoding)

    evt.next()

File: plugins/doc.js
"use strict";

module.exports = {

    beeld_plugin_doc: function(beelder) {
        var Beeld = beelder.getClass(),
            HAS = Object.prototype.hasOwnProperty,
            write_async = Beeld.Utils.write_async,
            get_real_path = Beeld.Utils.get_real_path,
            regex = Beeld.Utils.regex,
            xpresion = Beeld.Utils.xpresion,
            evaluate = Beeld.Utils.evaluate
        ;
        var action_doc = function action_doc(evt) {
            var params = evt.data,
                options = params.options,
                data = params.data,
                current = params.current,
                doc = current.action_cfg;
            if (doc && HAS.call(doc,'output'))
            {
                doc['output'] = xpresion(doc['output'], evt); // parse xpresion if any
                var docFile = get_real_path(evaluate(doc['output'], {}), options.basePath),
                    docs = [], startDoc = doc['startdoc'], endDoc = doc['enddoc'],
                    isRegex = 0, _trim = null, _trimlen = 0,
                    blocks, i, l, tmp, j, l2, sep
                ;

                sep = HAS.call(doc,'separator') ? doc['separator'] : "\n\n";

                if (HAS.call(doc,'trim'))
                {
                    _trim = regex(doc.trim, evt);
                    if (false === _trim)
                    {
                        _trim = doc.trim;
                        _trimlen = _trim.length;
                        isRegex = 0;
                    }
                    else
                    {
                        isRegex = 1;
                    }
                }

                // extract doc blocks
                blocks = data.src.split(startDoc);
                l = blocks.length;
                for (i=0; i<l; ++i)
                {
                    tmp = blocks[i].split(endDoc);
                    if (tmp.length > 1) docs.push(tmp[0]);
                }
                blocks = null;

                // trim start of each doc block line
                if (_trim)
                {
                    l = docs.length;
                    for (i=0; i<l; ++i)
                    {
                        tmp = docs[i].split("\n");
                        l2 = tmp.length;
                        for (j=0; j<l2; ++j)
                        {
                            if (tmp[j].length)
                            {
                                if (isRegex)
                                {
                                    tmp[j] = tmp[j].replace(_trim, '');
                                }
                                else if (_trim === tmp[j].substr(0, _trimlen))
                                {
                                    tmp[j] = tmp[j].substr(_trimlen);
                                }
                            }
                        }
                        docs[i] = tmp.join("\n");
                    }
                }
                write_async(docFile, docs.join(sep), options.encoding);
            }
            evt.next();
        };

        beelder.addAction('doc', action_doc);
    }
};
